### PR TITLE
Never let cluster-scoped resources skip webhooks

### LINKED ISSUE #57964: admission webhooks: broken for cluster-scoped resources
If I am not mistaken, https://github.com/kubernetes/kubernetes/blob/7945ae68d0c7cffb070d60ad4d8bfe6ef585c279/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/namespace/matcher.go#L96 means that cluster-scoped resources other than namespaces are broken with admission webhooks.

There is even a unit test checking that: https://github.com/kubernetes/kubernetes/blob/7945ae68d0c7cffb070d60ad4d8bfe6ef585c279/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/namespace/matcher_test.go#L116.

### PR DESCRIPTION
**Pull request by caesarxuchao**
Fix #57964

This allows user write webhooks for cluster-scoped custom resources.

We still need to figure out how to selectively exempt cluster-scoped resources from webhooks to avoid bootstrapping deadlocks. For now, if a deadlock occurs, users can work around by first deleting the webhook configuration, then rebooting the webhook, then re-enabling the webhook configuration.

```release-note
Bug fix: webhooks now do not skip cluster-scoped resources

Action required: Before upgrading your Kubernetes clusters, double check if you had configured webhooks for cluster-scoped objects (e.g., nodes, persistentVolume), these webhooks will start to take effect. Delete/modify the configs if that's not desirable.
```

### CODE REVIEW

**Comment by sttts** [comment]
Please update also the type docs https://github.com/sttts/kubernetes/blob/951962512b9cfe15b25e9c715a5f33f088854f97/staging/src/k8s.io/api/admissionregistration/v1beta1/types.go#L150. Then lgtm.

File: pkg/apis/admissionregistration/types.go [hunk 1]
```diff
@@ -196,7 +196,7 @@ type Webhook struct {
 	// on whether the namespace for that object matches the selector. If the
 	// object itself is a namespace, the matching is performed on
 	// object.metadata.labels. If the object is other cluster scoped resource,
-	// it is not subjected to the webhook.
+	// it never skips the webhook.
```
Comment (caesarxuchao) [inline pkg/apis/admissionregistration/types.go]: @sttts PTAL. Sorry it has been off my radar for a while.

Comment (sttts) [reply pkg/apis/admissionregistration/types.go]: nit: another cluster scoped resource.

But I am no native speaker either @deads2k you are. Can you comment?

**Comment by sttts** [comment]
lgtm modulo the nit.

/approve

File: pkg/apis/admissionregistration/types.go (hunk 1 above)
Comment (caesarxuchao) [reply pkg/apis/admissionregistration/types.go]: Confirmed with @cheftako that we should use "another" :)

**Comment by caesarxuchao** [comment]
Nit fixed. I'll self apply the lgtm.

File: pkg/apis/admissionregistration/types.go (hunk 1 above)
Comment (sttts) [reply pkg/apis/admissionregistration/types.go]: :)

**Comment by sttts** [comment]
/assign @lavalamp 

For the api Godoc changes.

**Comment by lavalamp** [comment]
This is a breaking API change. I think at a minimum it needs an "action required" notice in the release notes. Anyone who has added a webhook for a cluster scoped resource and forgotten about it is going to be in for a surprise when they get this patch.

I think I agree that this is a bug that should be fixed, but is it so urgent that we want to do it before adding some exception mechanism?

**Comment by deads2k** [comment]
> This is a breaking API change. I think at a minimum it needs an "action required" notice in the release notes. Anyone who has added a webhook for a cluster scoped resource and forgotten about it is going to be in for a surprise when they get this patch.

No issue with adding "action required", but worrying about a case where someone created the config, tested it, realized it didn't work, didn't remove it, upgrades, and then gets upset that it starts working seems like an edge of an edge of an edge for a config that is already broken.



> I think I agree that this is a bug that should be fixed, but is it so urgent that we want to do it before adding some exception mechanism?

Adding an exception mechanism would be a backward compatible change.  Having exceptions on cluster-scoped things is less obvious of a use-case since they're already privileged so there are likely to be fewer webhooks for them.

I had thought about applying the same label selection scheme since the data is already present (labels on the object you're touching or its previous value), but I wouldn't have blocked this on it.

**Comment by lavalamp** [comment]
I'm the last person to apply rules for their own sake, and I agree this should be very rare, but I still worry about whether people will trust our API stability guarantees when we're willing to break things like this. I won't block it if we add the action-required notice.

I should write a blog post called "it's not safe to upgrade your cluster".

**Comment by caesarxuchao** [comment]
Added "action required". Feel free to reword it.

**Pull request merged**